File: javascript/jssearch/src/searcher.js
/*
 * searcher.js
 *
 * performs the searching based on the given SearchSettings instance
 */

const assert = require('assert');
const {common, FileType, FileUtil, Finder} = require('jsfind');
const {SearchError} = require('./searcherror');
const {SearchResult} = require('./searchresult');
const path = require('path')

class Searcher {
    'use strict'

    constructor(settings) {
        this.settings = settings;
        this.binaryEncoding = 'latin1';
        // from https://github.com/nodejs/node/blob/master/lib/buffer.js
        this.supportedEncodings = ['utf-8', 'utf8', 'latin1', 'ascii', 'ucs2',  'ucs-2', 'utf16le',
            'binary', 'base64', 'hex'];
        this.finder = new Finder(settings.findSettings);
        this.validateSettings();
    }

    validateSettings() {
        try {
            assert.ok(this.settings.searchPatterns.length, 'No search patterns defined');
            assert.ok(this.supportedEncodings.indexOf(this.settings.textFileEncoding) > -1,
                'Invalid encoding');
            assert.ok(this.settings.linesBefore > -1, 'Invalid linesbefore');
            assert.ok(this.settings.linesAfter > -1, 'Invalid linesafter');
            assert.ok(this.settings.maxLineLength > -1, 'Invalid maxlinelength');

        } catch (err) {
            let msg = err.message;
            if (err.code === 'ENOENT') {
                msg = 'Startpath not found';
            } else if (err.code === 'EACCES') {
                msg = 'Startpath not readable';
            }
            throw new SearchError(msg);
        }
    }

    matchesAnyElement(s, elements) {
        return elements.indexOf(s) > -1;
    }

    matchesAnyPattern(s, patterns) {
        return patterns.some((p, i, arr) => s.search(p) > -1);
    }

    anyMatchesAnyPattern(ss, patterns) {
        return ss.some((s, i, arr) => this.matchesAnyPattern(s, patterns));
    }

    async search() {
        try {
            // get the search files
            let fileResults = await this.finder.find();
    
            if (this.settings.verbose) {
                let dirs = fileResults.map(sf => sf.path);
                dirs = common.setFromArray(dirs);
                dirs.sort();
                common.log("\nDirectories to be searched " + `(${dirs.length}):`);
                dirs.forEach(d => common.log(d));

                common.log("\nFiles to be searched " + `(${fileResults.length}):`);
                fileResults.forEach(sf => common.log(sf.relativePath()));
                common.log("");
            }

            // search the files
            let results = [];
            const searchResultsArrays = await Promise.all(fileResults.map(fr => this.searchFile(fr)));
            searchResultsArrays.forEach(searchResults => {
                results = results.concat(searchResults);
            });

            if (this.settings.verbose) {
                common.log('Search complete.');
            }

            return results;

        } catch (err) {
            throw err;
        }
    }

    async searchFile(fileResult) {
        let results = [];
        switch (fileResult.fileType) {
            case FileType.CODE:
            case FileType.TEXT:
            case FileType.XML:
                results = await this.searchTextFile(fileResult);
                break;
            case FileType.BINARY:
                results = await this.searchBinaryFile(fileResult);
                break;
            default:
                // TODO: add message about unsupported fileType
                break;
        }
        return results;
    }

    async searchBinaryFile(fileResult) {
        if (this.settings.verbose) {
            common.log(`Searching binary file: "${fileResult}"`);
        }

        const contents = FileUtil.getFileContentsSync(fileResult.relativePath(), this.binaryEncoding);
        let results = [];

        const searchPattern = p => {
            let pattern = new RegExp(p.source, 'g');
            let patternResults = [];
            let match = pattern.exec(contents);
            while (match) {
                patternResults.push(new SearchResult(
                    pattern,
                    fileResult,
                    0,
                    match.index+1,
                    pattern.lastIndex+1,
                    null,
                    [],
                    []));
                if (this.settings.firstMatch) {
                    return patternResults;
                }
                match = pattern.exec(contents);
            }
            return patternResults;
        }

        const patternResultArrays = await Promise.all(this.settings.searchPatterns.map(p => searchPattern(p)));
        patternResultArrays.forEach(patternResults => {
            results = results.concat(patternResults);
        });
        return results;
    }

    async searchTextFile(fileResult) {
        if (this.settings.verbose) {
            common.log(`Searching text file ${fileResult}`);
        }
        let results;
        if (this.settings.multilineSearch) {
            results = await this.searchTextFileContents(fileResult);
        } else {
            results = await this.searchTextFileLines(fileResult);
        }
        return results;
    }

    async searchTextFileContents(fileResult) {
        const contents = FileUtil.getFileContentsSync(fileResult.relativePath(), this.settings.textFileEncoding);
        let stringResults = await this.searchMultiLineString(contents);
        return stringResults.map(r => {
            return new SearchResult(r.pattern, fileResult, r.lineNum, r.matchStartIndex, r.matchEndIndex, r.line,
                r.linesBefore, r.linesAfter, this.settings.maxLineLength, this.settings.colorize);
        });
    }

    getNewLineIndices(s) {
        let indices = [];
        for (let i = 0; i < s.length; i++) {
            if (s.charAt(i) === "\n") {
                indices.push(i);
            }
        }
        return indices;
    }

    getLinesAtIndices(s, atIndices, startLineIndices, endLineIndices) {
        if (atIndices.length === 0)
            return [];
        let lines = [];
        atIndices.forEach(i => {
            let line = s.substring(i, endLineIndices[startLineIndices.indexOf(i)]);
            lines.push(line);
        });
        return lines;
    }

    getLinesBefore(s, beforeStartIndices, startLineIndices, endLineIndices) {
        return this.getLinesAtIndices(s, beforeStartIndices, startLineIndices, endLineIndices);
    }

    getLinesAfter(s, afterStartIndices, startLineIndices, endLineIndices) {
        return this.getLinesAtIndices(s, afterStartIndices, startLineIndices, endLineIndices);
    }

    getLessThanOrEqual(matchVal) {
        return i => { return i <= matchVal; };
    }

    getGreaterThan(matchVal) {
        return i => { return i > matchVal; };
    }

    plusOne(i) {
        return i + 1;
    }

    async searchMultiLineString(s) {
        let linesBefore = [];
        let linesAfter = [];
        let results = [];
        try {
            let newLineIndices = this.getNewLineIndices(s);
            let startLineIndices = [0].concat(newLineIndices.map(this.plusOne));
            let endLineIndices = newLineIndices.concat([s.length - 1]);

            const searchPattern = pattern => {
                pattern = new RegExp(pattern.source, 'g');
                let patternResults = [];
                let match = pattern.exec(s);
                while (match) {
                    if (this.settings.firstMatch && patternResults.length > 0) {
                        return patternResults;
                    }
                    let lessOrEqual = this.getLessThanOrEqual(match.index);
                    let greaterThan = this.getGreaterThan(match.index);
                    let lineStartIndex = 0;
                    let lineEndIndex = s.length - 1;
                    let beforeLineCount = 0;
                    let beforeStartIndices = startLineIndices.filter(lessOrEqual);
                    if (beforeStartIndices.length > 0) {
                        lineStartIndex = beforeStartIndices.pop();
                        beforeLineCount = beforeStartIndices.length;
                        if (beforeStartIndices.length > this.settings.linesBefore) {
                            beforeStartIndices = beforeStartIndices.slice(
                                beforeStartIndices.length - this.settings.linesBefore);
                        }
                    }
                    lineEndIndex = endLineIndices[startLineIndices.indexOf(lineStartIndex)];
                    let line = s.substring(lineStartIndex, lineEndIndex);
                    if (this.settings.linesBefore && beforeLineCount) {
                        linesBefore = this.getLinesBefore(s, beforeStartIndices,
                            startLineIndices, endLineIndices);
                    }
                    if (this.settings.linesAfter) {
                        let afterStartIndices = startLineIndices.filter(greaterThan);
                        if (afterStartIndices.length > this.settings.linesAfter) {
                            afterStartIndices = afterStartIndices.slice(0,
                                this.settings.linesAfter);
                        }
                        linesAfter = this.getLinesAfter(s, afterStartIndices,
                            startLineIndices, endLineIndices);
                    }
                    let matchStartIndex = match.index - lineStartIndex + 1;
                    let matchEndIndex = pattern.lastIndex - lineStartIndex + 1;
                    if ((this.settings.linesBefore === 0 || this.linesBeforeMatch(linesBefore)) &&
                        (this.settings.linesAfter === 0 || this.linesAfterMatch(linesAfter))) {
                        patternResults.push(new SearchResult(
                            pattern,
                            null,
                            beforeLineCount + 1,
                            matchStartIndex,
                            matchEndIndex,
                            line,
                            [].concat(linesBefore),
                            [].concat(linesAfter),
                            this.settings.maxLineLength,
                            this.settings.colorize));
                        if (!(pattern.source in patternResults)) {
                            patternResults[pattern.source] = 1;
                        }
                    }
                    match = pattern.exec(s);
                }
                return patternResults;
            }

            const patternResultArrays = await Promise.all(this.settings.searchPatterns.map(p => searchPattern(p)));
            patternResultArrays.forEach(patternResults => {
                results = results.concat(patternResults);
            });
            return results;

        } catch (err) {
            throw err;
        }
    }

    linesMatch(lines, inPatterns, outPatterns) {
        return ((inPatterns.length === 0 || this.anyMatchesAnyPattern(lines, inPatterns)) &&
            (outPatterns.length === 0 || ! this.anyMatchesAnyPattern(lines, outPatterns)));
    }

    linesBeforeMatch(linesBefore) {
        return this.linesMatch(linesBefore, this.settings.inLinesBeforePatterns,
            this.settings.outLinesBeforePatterns);
    }

    linesAfterMatch(linesAfter) {
        return this.linesMatch(linesAfter, this.settings.inLinesAfterPatterns,
            this.settings.outLinesAfterPatterns);
    }

    async searchTextFileLines(fileResult) {
        let lines = FileUtil.getFileLinesSync(fileResult.relativePath(), this.settings.textFileEncoding);
        let linesResults = await this.searchLines(lines);
        return linesResults.map(r => {
            return new SearchResult(r.pattern, fileResult, r.lineNum, r.matchStartIndex, r.matchEndIndex, r.line,
                r.linesBefore, r.linesAfter, this.settings.maxLineLength, this.settings.colorize);
        });
    }

    // return results so that filepath can be added to them
    async searchLines(lines) {
        let lineNum = 0;
        let pattern;
        let linesBefore = [];
        let linesAfter = [];
        let results = [];
        let patternResults = {};
        while (true) {
            if (Object.keys(patternResults).length === this.settings.searchPatterns.length) {
                break;
            }
            let line = "";
            if (linesAfter.length > 0) {
                line = linesAfter.shift();
            } else if (lines.length > 0) {
                line = lines.shift();
            } else {
                break;
            }
            lineNum += 1;
            if (this.settings.linesAfter > 0) {
                while (linesAfter.length < this.settings.linesAfter && lines.length > 0) {
                    linesAfter.push(lines.shift());
                }
            }
            this.settings.searchPatterns.forEach(p => {
                pattern = new RegExp(p.source, "g");
                let match = pattern.exec(line);
                while (match) {
                    if ((this.settings.linesBefore === 0 || this.linesBeforeMatch(linesBefore)) &&
                        (this.settings.linesAfter === 0 || this.linesAfterMatch(linesAfter))) {
                        results.push(new SearchResult(
                            pattern,
                            null,
                            lineNum,
                            match.index+1,
                            pattern.lastIndex+1,
                            line,
                            [...linesBefore],
                            [...linesAfter],
                            this.settings.maxLineLength,
                            this.settings.colorize));
                        if (this.settings.firstMatch) {
                            patternResults[pattern.source] = 1;
                            break;
                        }
                    }
                    match = pattern.exec(line);
                }
            });
            if (this.settings.linesBefore > 0) {
                if (linesBefore.length === this.settings.linesBefore)
                    linesBefore.shift();
                if (linesBefore.length < this.settings.linesBefore)
                    linesBefore.push(line);
            }
        }
        return results;
    }

    cmpSearchResults(r1, r2) {
        const pathCmp = r1.file.path.localeCompare(r2.file.path);
        if (pathCmp === 0) {
            const fileCmp = path.basename(r1.file.fileName).localeCompare(path.basename(r2.file.fileName));
            if (fileCmp === 0) {
                if (r1.lineNum === r2.lineNum) {
                    return r1.matchStartIndex - r2.matchStartIndex;
                }
                return r1.lineNum - r2.lineNum;
            }
            return fileCmp;
        }
        return pathCmp;
    }

    printSearchResults(results, formatter) {
        // first sort the results
        // TODO: ensure sorting gets help from jsfind
        results.sort(this.cmpSearchResults);
        common.log(`\nSearch results (${results.length}):`);
        results.forEach(r => common.log(formatter.format(r)));
    }

    getFileResults(results) {
        let fileMap = {};
        let fileResults = [];
        for (let r of results) {
            if (!fileMap[r.file.relativePath()]) {
                fileMap[r.file.relativePath()] = r.file;
                fileResults.push(r.file);
            }
        }
        return fileResults;
    }

    printMatchingDirs(results, formatter) {
        const fileResults = this.getFileResults(results);
        this.finder.printMatchingDirs(fileResults, formatter.fileFormatter);
    }

    printMatchingFiles(results, formatter) {
        const fileResults = this.getFileResults(results);
        this.finder.printMatchingFiles(fileResults, formatter.fileFormatter);
    }

    getMatchingLines(results) {
        let lines = results.filter(r => r.lineNum > 0).map(r => r.line.trim());
        if (this.settings.uniqueLines) {
            lines = common.setFromArray(lines);
        }
        lines.sort((a, b) => {
            let aUpper = a.toUpperCase();
            let bUpper = b.toUpperCase();
            if (aUpper === bUpper)
                return 0;
            return aUpper < bUpper ? -1 : 1;
        });
        return lines;
    }

    printMatchingLines(results, formatter) {
        const lines = this.getMatchingLines(results);
        let hdrText;
        if (this.settings.uniqueLines)
            hdrText = `\nUnique lines with matches`;
        else
            hdrText = `\nLines with matches`;
        if (lines.length > 0) {
            hdrText = `${hdrText} (${lines.length}):`;
            common.log(hdrText);
            lines.forEach(l => common.log(formatter.formatLine(l)));
        } else {
            hdrText = `${hdrText}: 0`;
            common.log(hdrText);
        }
    }
}

exports.Searcher = Searcher;

File: elixir/exsearch/lib/searcher.ex
defmodule ExSearch.Searcher do
  @moduledoc """
  Documentation for `ExSearch.Searcher`.
  """

  alias ExFind.FileResult
  alias ExFind.Finder
  alias ExFind.StringUtil

  alias ExSearch.SearchError
  alias ExSearch.SearchResult
  alias ExSearch.SearchSettings

  defstruct [:settings]

  def new(settings) do
    __struct__([settings: settings])
  end

  # ----------------------------------------------------------------------------
  # Binary search (blob search)
  # ----------------------------------------------------------------------------
  def blob_match_index_to_search_result(pattern, match_index) do
    {match_idx, match_length} = match_index
    SearchResult.new(pattern, nil, 0, "", [], [], match_idx + 1, match_idx + match_length + 1)
  end

  def search_blob_for_pattern(searcher, blob, pattern) do
    matching_indices =
      if searcher.settings.first_match do
        Regex.run(pattern, blob, return: :index)
      else
        Regex.scan(pattern, blob, return: :index) |> Enum.map(fn i -> List.first(i) end)
      end
    case matching_indices do
      nil -> []
      [] -> []
      _ ->
        Enum.map(matching_indices, fn i -> blob_match_index_to_search_result(pattern, i) end)
    end
  end

  def search_blob(searcher, blob) do
    searcher.settings.search_patterns
    |> Enum.map(fn p -> search_blob_for_pattern(searcher, blob, p) end)
    |> List.flatten()
  end

  def search_binary_file(searcher, file_result) do
    case File.open(FileResult.relative_path(file_result), [:binary, :read]) do
      {:error, message} -> {:error, message}
      {:ok, file} ->
        search_results = search_blob(searcher, IO.binread(file, :eof))
        Enum.map(search_results, fn sr -> %{sr | file: file_result} end)
    end
  end

  # ----------------------------------------------------------------------------
  # Text-based search (multi-line search)
  #
  # This search method is extremely slow and I haven't pinpointed the reason yet,
  # so for now it's disabled.
  # ----------------------------------------------------------------------------
  def text_match_index_to_search_result(searcher, text, pattern, match_index, start_line_indices, end_line_indices) do
    {match_idx, match_length} = match_index
    prev_start_line_indices = Enum.take_while(start_line_indices, fn i -> i < match_idx end)
    line_num = length(prev_start_line_indices)
    prev_end_line_indices = Enum.take(end_line_indices, line_num)
    line_start_index = List.last(prev_start_line_indices)
    line_end_index = List.last(prev_end_line_indices)
    line = String.slice(text, line_start_index, line_end_index - line_start_index)
    lines_before =
      if searcher.settings.lines_before > 0 do
        before_start_indices = Enum.drop(prev_start_line_indices, -1)
                               |> Enum.take(searcher.settings.lines_before * -1)
        before_end_indices = Enum.drop(prev_end_line_indices, -1)
                             |> Enum.take(searcher.settings.lines_before * -1)
        Enum.zip(before_start_indices, before_end_indices)
        |> Enum.map(fn {s, e} -> String.slice(text, s, e - s) end)
      else
        []
      end
    lines_after =
      if searcher.settings.lines_after > 0 do
        after_start_indices = Enum.drop(start_line_indices, line_num)
                              |> Enum.take(searcher.settings.lines_after)
        after_end_indices = Enum.drop(end_line_indices, line_num)
                            |> Enum.take(searcher.settings.lines_after)
        Enum.zip(after_start_indices, after_end_indices)
        |> Enum.map(fn {s, e} -> String.slice(text, s, e - s) end)
      else
        []
      end
    SearchResult.new(pattern, nil, line_num, line, lines_before, lines_after, match_idx - line_start_index + 1, match_idx - line_start_index + match_length + 1)
  end

  def search_text_for_pattern(searcher, text, pattern, start_line_indices, end_line_indices) do
    matching_indices =
      if searcher.settings.first_match do
        Regex.run(pattern, text, return: :index)
      else
        Regex.scan(pattern, text, return: :index) |> Enum.map(fn i -> List.first(i) end)
      end
    if matching_indices != nil and matching_indices != [] do
      IO.puts("matching_indices: #{inspect(matching_indices)}")
    end
    case matching_indices do
      nil -> []
      [] -> []
      _ ->
        Enum.map(matching_indices, fn i -> text_match_index_to_search_result(searcher, text, pattern, i, start_line_indices, end_line_indices) end)
    end
  end

  def search_text(searcher, text) do
    charlist = String.to_charlist(text)
    new_line_indices = Enum.to_list(0..length(charlist) - 1)
                       |> Enum.reduce([], fn i, acc -> if Enum.at(charlist, i) == 10, do: acc ++ [i], else: acc end)
    start_line_indices = [0] ++ Enum.map(new_line_indices, fn i -> i + 1 end)
    end_line_indices = new_line_indices ++ [String.length(text)]
    searcher.settings.search_patterns
    |> Enum.map(fn p -> search_text_for_pattern(searcher, text, p, start_line_indices, end_line_indices) end)
    |> List.flatten()
  end

  def search_text_file_contents(searcher, file_result) do
    case File.open(FileResult.relative_path(file_result), [{:encoding, searcher.settings.text_file_encoding}]) do
      {:error, message} -> {:error, message}
      {:ok, file} ->
        search_results = search_text(searcher, IO.read(file, :eof))
        Enum.map(search_results, fn sr -> %{sr | file: file_result} end)
    end
  end

  # ----------------------------------------------------------------------------
  # Line-based search
  # ----------------------------------------------------------------------------
  def line_match_index_to_search_result(line_num, line, lines_before, lines_after, pattern, match_index) do
    {match_idx, match_length} = match_index
    SearchResult.new(pattern, nil, line_num, String.trim_trailing(line), Enum.map(lines_before, &(String.trim_trailing(&1))),
                     Enum.map(lines_after, &(String.trim_trailing(&1))), match_idx + 1, match_idx + match_length + 1)
  end

  def search_line_for_pattern(searcher, line_num, line, pattern, lines_before, lines_after) do
    # first filter on lines_before and lines_after
    if (!Enum.empty?(searcher.settings.in_lines_before_patterns)
        and !StringUtil.any_matches_any_pattern(lines_before, searcher.settings.in_lines_before_patterns))
        or (!Enum.empty?(searcher.settings.out_lines_before_patterns)
            and StringUtil.any_matches_any_pattern(lines_before, searcher.settings.out_lines_before_patterns))
        or (!Enum.empty?(searcher.settings.in_lines_after_patterns)
            and !StringUtil.any_matches_any_pattern(lines_after, searcher.settings.in_lines_after_patterns))
        or (!Enum.empty?(searcher.settings.out_lines_after_patterns)
            and StringUtil.any_matches_any_pattern(lines_after, searcher.settings.out_lines_after_patterns)) do
      []
    else
      matching_indices =
        if searcher.settings.first_match do
          Regex.run(pattern, line, return: :index)
        else
          Regex.scan(pattern, line, return: :index) |> Enum.map(fn i -> List.first(i) end)
        end
      case matching_indices do
        nil -> []
        [] -> []
        _ ->
          Enum.map(matching_indices, fn i -> line_match_index_to_search_result(line_num, line, lines_before, lines_after, pattern, i) end)
      end
    end
  end

  def line_stream_to_results(searcher, line_stream, lines_before, results) do
    if searcher.settings.first_match and results != [] do
      results
    else
      case Enum.take(line_stream, 1) do
        [] -> results
        [{line, line_num}] ->
          lines_after =
            if searcher.settings.lines_after > 0 do
              Enum.drop(line_stream, 1)
              |> Enum.take(searcher.settings.lines_after)
              |> Enum.map(fn {l, _ln} -> l end)
            else
              []
            end
          line_results =
            searcher.settings.search_patterns
            |> Enum.map(fn p -> search_line_for_pattern(searcher, line_num, line, p, lines_before, lines_after) end)
            |> List.flatten()
          new_lines_before =
            if searcher.settings.lines_before > 0 do
              if length(lines_before) < searcher.settings.lines_before do
                lines_before ++ [line]
              else
                Enum.drop(lines_before, 1) ++ [line]
              end
            else
              []
            end
          line_stream_to_results(searcher, Enum.drop(line_stream, 1), new_lines_before, results ++ line_results)
      end
    end
  end

  def search_line_stream(searcher, stream) do
    line_stream =
      stream
      |> Stream.with_index(1)
   line_stream_to_results(searcher, line_stream, [], [])
  end

  def search_text_file_lines(searcher, file_result) do
    search_results =
      search_line_stream(searcher, File.stream!(FileResult.relative_path(file_result), [encoding: searcher.settings.text_file_encoding]))
      |> Enum.to_list()
    Enum.map(search_results, fn sr -> %{sr | file: file_result} end)
  end

  def search_text_file(searcher, file_result) do
    if searcher.settings.multi_line_search do
      # Contents searching is disabled for now, too slow
      # search_text_file_contents(searcher, file_result)
      search_text_file_lines(searcher, file_result)
    else
      search_text_file_lines(searcher, file_result)
    end
  end

  def search_file(searcher, file_result) do
    case file_result.file_type do
      :archive -> []
      :binary -> search_binary_file(searcher, file_result)
      t ->
        if Enum.member?([:code, :xml, :text], t), do: search_text_file(searcher, file_result), else: []
    end
  end

  def search_files(searcher, file_results) do
    search_results = Enum.map(file_results, fn fr -> search_file(searcher, fr) end)
    |> List.flatten()
    {:ok, search_results}
  end

  def search_files_async(searcher, file_results) do
    tasks = Enum.map(file_results, fn fr -> Task.async(fn -> search_file(searcher, fr) end) end)
    search_results = Task.await_many(tasks) |> List.flatten()
    {:ok, search_results}
  end

  def search(searcher) do
    # temporary for turning on/off async search
    do_async = true
    case validate_settings(searcher.settings) do
      {:error, message} -> {:error, message}
      {:ok, _} ->
        find_settings = SearchSettings.to_find_settings(searcher.settings)
        finder = Finder.new(find_settings)
        case Finder.find(finder) do
          {:error, message} -> {:error, message}
          {:ok, file_results} ->
            if do_async do
              search_files_async(searcher, file_results)
            else
              search_files(searcher, file_results)
            end
        end
    end
  end

  def search!(searcher) do
    case search(searcher) do
      {:error, message} -> raise SearchError, message: message
      {:ok, results} -> results
    end
  end

  def validate_settings(settings) do
    case Finder.validate_settings(settings) do
      {:error, message} -> {:error, message}
      {:ok, _} ->
        cond do
          Enum.empty?(settings.search_patterns) ->
            {:error, "No search patterns defined"}
          settings.lines_after < 0 ->
            {:error, "Invalid linesafter"}
          settings.lines_before < 0 ->
            {:error, "Invalid linesbefore"}
          settings.max_line_length < 0 ->
            {:error, "Invalid maxlinelength"}
          true -> {:ok, "Settings are valid"}
        end
    end
  end

end
